File: cpp/123.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  int maxProfit(vector<int> &prices) {
    int n = prices.size();
    if(n <= 1) return 0;
    int mi = prices[0], mx = -prices[0];
    int once[n], twice[n];
    once[0] = twice[0] = 0;
    for(int i=1; i<n; i++){
      once[i] = max(once[i-1], prices[i] - mi);
      twice[i] = max(twice[i-1], mx + prices[i]);
      cout << once[i] << " " << twice[i] << endl;
      mi = min(mi, prices[i]);
      mx = max(mx, once[i-1] - prices[i]);
    }
    return max(once[n-1], twice[n-1]);
  }
};

int main(){
  auto sol = Solution();
  vector<int> v = {2,1,4,5,2,9,7};
  cout << sol.maxProfit(v);
}

File: python/123.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        if not prices:
            return 0
        mi = prices[0]
        mx1 = 0
        mx2 = -prices[0]
        ans = 0
        for p in prices[1:]:
            ans = max(ans, p + mx2)
            mx2 = max(mx2, mx1 - p)
            mx1 = max(mx1, p - mi)
            mi = min(mi, p)
        return ans
